feat: Add Copilot agent instructions (migrated from OMEGA) (#35322)
openvinotoolkit/openvino: .github/agents-prototype/skills/add-fe-op/onnx.md

# Skill: Add a New Operation Translator to the OpenVINO ONNX Frontend

---

## 1. Prerequisites

Before writing any code:

- [ ] Read the [ONNX spec for the op](https://onnx.ai/onnx/operators/) — inputs, outputs, attributes, type constraints, and behavioral notes.
- [ ] Check the [ONNX Operator Changelog](https://github.com/onnx/onnx/blob/main/docs/Changelog.md) for all opset versions that introduced or modified the op.
- [ ] Identify the matching OpenVINO op(s) in `src/core/include/openvino/op/`. If none exists, use a composition of existing ops or escalate to the Core OpSpec Agent.
- [ ] Search `src/frontends/common_translators/` — a shared translator may already exist or be partially implemented.
- [ ] Check if other frontends (PyTorch, TF) already have a translator for the same operation; it may be reusable or a model to follow.

---

## 2. Key Source Locations

| Path | Description |
|---|---|
| `src/frontends/onnx/frontend/src/op/` | Per-op translator files — create your new file here. |

> | `src/frontends/onnx/frontend/src/ops_bridge.cpp` | Operator registry — add `ONNX_OP` registration here. |

Reviewer: This info looks deprecated (~1 year) - the ONNX_OP registration should be registered in the ONNX op translation file.
The ops_bridge.cpp is no longer the place to put ONNX_OP registration.
https://github.com/openvinotoolkit/openvino/blob/52881e1528fc5f2b108ff2e84205cc769dc18c8f/src/frontends/onnx/frontend/src/ops_bridge.cpp

Author: @mvafin for your attention

> | `src/frontends/onnx/frontend/src/ops_bridge.cpp` | Operator registry — add `ONNX_OP` registration here. |

Reviewer: ```suggestion
| `src/frontends/onnx/frontend/src/op/` | Per-op translator files — create your new file here with `ONNX_OP` registration.|
| `src/frontends/onnx/frontend/src/ops_bridge.cpp` | Operator registry definition. |
```

std::string auto_pad = node.get_attribute_value<std::string>("auto_pad", "NOTSET");

// Check if an attribute exists:
bool has_pads = node.has_attribute("pads");

// Check if an optional input is present (not "missing"):
#include "utils/common.hpp"
bool valid = common::is_input_valid(node, 2);
```

---

## 4. Registration Macros

In `ops_bridge.cpp`:

```cpp
#include "op/<new_op>.hpp"

namespace ai_onnx {
namespace opset_1 {
// Define translator in .hpp/.cpp file
}  // namespace opset_1
}  // namespace ai_onnx

// Register with a single opset version range:
ONNX_OP("GridSample", OPSET_SINCE(16), ai_onnx::opset_16::grid_sample);

// Register with version ranges — when behavior changed at opset 11:
ONNX_OP("Resize", OPSET_RANGE(10, 10), ai_onnx::opset_10::resize);
ONNX_OP("Resize", OPSET_RANGE(11, 17), ai_onnx::opset_11::resize);
ONNX_OP("Resize", OPSET_SINCE(18), ai_onnx::opset_18::resize);

// Register for specific opset versions (when there are gaps):
ONNX_OP("Cast", OPSET_IN({1, 6, 9, 13, 19}), ai_onnx::opset_13::cast);
```

---

## 5. Step-by-Step Implementation

### 5a. Create the translator file

```bash
# Create the header file:

> cat > src/frontends/onnx/frontend/src/op/<new_op>.hpp << 'EOF'

Reviewer: This instruction to create .hpp file is not aligned with current code base. There are no op .hpp files for ONNX ops in the https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/onnx/frontend/src/op
or in other directory.

Reviewer 2: That is quite important, we really don't have .hpp files for operations, only .cpp files need to be created

Author: @mvafin - addressed

// Copyright (C) 2018-2026 Intel Corporation
// SPDX-License-Identifier: Apache-2.0

#pragma once

#include "core/node.hpp"

namespace ov {
namespace frontend {
namespace onnx {
namespace ai_onnx {

namespace opset_<first_version> {
ov::OutputVector <new_op>(const ov::frontend::onnx::Node& node);
}  // namespace opset_<first_version>

}  // namespace ai_onnx
}  // namespace onnx
}  // namespace frontend
}  // namespace ov

> EOF

Reviewer: Such strict instructions - like bash command to create a file and exact copy paste initial code seems too hardcoded and not fully benefit from the models intelligence - If some code refactor is applied to the project - the skills are instantly deprecated and require maintenance. Agents can easily retrive the current code base structure without it.

namespace onnx {
namespace ai_onnx {

namespace opset_<first_version> {

ov::OutputVector <new_op>(const ov::frontend::onnx::Node& node) {
    const auto inputs = node.get_ov_inputs();
    const auto x = inputs[0];

    // ... read attributes ...
    const auto axis = node.get_attribute_value<int64_t>("axis", 0);

    // ... build OpenVINO subgraph ...
    return {std::make_shared<ov::op::v0::...>(x)};
}

}  // namespace opset_<first_version>

}  // namespace ai_onnx
}  // namespace onnx
}  // namespace frontend
}  // namespace ov
EOF
```

### 5b. Handling multiple opset versions

When an op was added in opset 9 and modified in opset 13:

```cpp
// ---- opset_9 namespace ----
namespace opset_9 {
ov::OutputVector my_op(const ov::frontend::onnx::Node& node) {
    // opset 9 implementation
}
}  // namespace opset_9

// ---- opset_13 namespace ----
namespace opset_13 {
ov::OutputVector my_op(const ov::frontend::onnx::Node& node) {
    // opset 13 implementation (new behavior/attributes)
}
}  // namespace opset_13
```

In `ops_bridge.cpp`:
```cpp
ONNX_OP("MyOp", OPSET_RANGE(9, 12), ai_onnx::opset_9::my_op);
ONNX_OP("MyOp", OPSET_SINCE(13), ai_onnx::opset_13::my_op);

> ```

Reviewer: Not in ops_bridge.cpp but in the onnx op file
```suggestion
```cpp
ONNX_OP("MyOp", OPSET_RANGE(9, 12), ai_onnx::opset_9::my_op);
ONNX_OP("MyOp", OPSET_SINCE(13), ai_onnx::opset_13::my_op);
```
```

```cpp
#include "common_translators.hpp"

ov::OutputVector my_op(const ov::frontend::onnx::Node& node) {
    return common_translators::translate_<op>(node.get_ov_inputs(), attributes...);
}
```

### 5f. Mark operations for special types

For ops that process complex or quantized tensors:

```cpp
#include "openvino/frontend/complex_type_mark.hpp"

// In normalize-step MatcherPass (if needed):
auto complex_mark = std::make_shared<ComplexTypeMark>(input, input_et);
```

### 5g. Normalize-step transformations

For ops that require post-translation graph rewrites (e.g., resolving Mark operations or merging multi-op patterns):

1. Create a `MatcherPass` in `src/frontends/onnx/frontend/src/transforms/` (or `src/frontends/common_translators/` for shared transforms).
2. Register in `FrontEnd::normalize()` in `frontend.cpp`.

### 5h. Update `CMakeLists.txt`

In `src/frontends/onnx/frontend/CMakeLists.txt`:
```cmake
set(SOURCES
    ...
    src/op/<new_op>.cpp
    ...
)
```

### 5i. Register in `ops_bridge.cpp`

In `src/frontends/onnx/frontend/src/ops_bridge.cpp`:
```cpp
#include "op/<new_op>.hpp"
...
ONNX_OP("<OpName>", OPSET_SINCE(<first_opset>), ai_onnx::opset_<first_opset>::<new_op>);

> ```

Reviewer: The info about op registration is deprecated (pointing to incorrect file) and  at the same time repeated too many times. It's simple macro registration, I don't see a benefit from mentioning several times, especially if the approach from the instruction is not aligned with the code base.

```suggestion
```

    test_case.run();
}
```

### 6c. Add multiple test scenarios

Cover all important cases:
- Basic case with default attributes
- Non-default attribute values  
- Edge cases (empty tensor, scalar, dynamic rank)
- Multiple opset versions if behavior differs

---

## 7. Build and Test

```bash
# Build the ONNX frontend:
cmake --build build --target openvino_onnx_frontend -j$(nproc)

# Build and run tests:
cmake --build build --target ov_onnx_frontend_tests -j$(nproc)
cd build && ctest -R "ov_onnx_frontend_tests" --output-on-failure -j$(nproc)

# Run a specific test:
ctest -R "onnx_model_<op_name>" -V

# Apply clang-format:
cmake --build build --target clang_format_fix_all -j$(nproc)
```

---

## 8. Validation Checklist

- [ ] **ONNX spec consulted** — all attributes, input/output optionality, type constraints covered.
- [ ] **All opset versions covered** — `ONNX_OP` registrations span the correct version ranges.
- [ ] **Optional inputs guarded** — `common::is_input_valid(node, N)` used for all optional inputs.
- [ ] **Default attribute values** — per ONNX spec for each opset version.
- [ ] **CMakeLists.txt updated** — new `.cpp` added to `SOURCES`.

> - [ ] **`ops_bridge.cpp` updated** — `ONNX_OP` registration added.

Reviewer: As previously mentioned - ops_bridge.cpp is not the right place to put ONNX_OP registration.
```suggestion
- [ ] `ONNX_OP` registration added.
```
